src/plugins/dashboard/public/attribute_service/attribute_service.tsx

```diff
@@ -34,75 +38,104 @@ import {
 } from '../../../../core/public';
 import {
   SavedObjectSaveModal,
-  showSaveModal,
   OnSaveProps,
   SaveResult,
   checkForDuplicateTitle,
 } from '../../../saved_objects/public';
-import {
-  EmbeddableStart,
-  EmbeddableFactory,
-  EmbeddableFactoryNotFoundError,
-  Container,
-} from '../../../embeddable/public';
 
 /**
  * The attribute service is a shared, generic service that embeddables can use to provide the functionality
  * required to fulfill the requirements of the ReferenceOrValueEmbeddable interface. The attribute_service
  * can also be used as a higher level wrapper to transform an embeddable input shape that references a saved object
  * into an embeddable input shape that contains that saved object's attributes by value.
  */
+export const ATTRIBUTE_SERVICE_DEFAULT_KEY = 'attributes';
+
+export interface AttributeServiceOptions<
+  A extends { title: string },
+  K extends string = typeof ATTRIBUTE_SERVICE_DEFAULT_KEY
+> {
+  attributesKey?: K;
+  customSaveMethod?: (
+    type: string,
+    attributes: A,
+    savedObjectId?: string
+  ) => Promise<{ id: string }>;
+  customUnwrapMethod?: (savedObject: SimpleSavedObject<A>) => A;
+}
+
 export class AttributeService<
   SavedObjectAttributes extends { title: string },
-  ValType extends EmbeddableInput & { attributes: SavedObjectAttributes },
-  RefType extends SavedObjectEmbeddableInput
+  AttributesKey extends string = typeof ATTRIBUTE_SERVICE_DEFAULT_KEY,
+  ValType extends EmbeddableInput &
+    { [key in AttributesKey]: SavedObjectAttributes } = EmbeddableInput &
+    { [key in AttributesKey]: SavedObjectAttributes },
+  RefType extends SavedObjectEmbeddableInput = SavedObjectEmbeddableInput
 > {
-  private embeddableFactory: EmbeddableFactory;
+  private embeddableFactory?: EmbeddableFactory;
+  private attributesKey: AttributesKey;
```

Comment:
Do we really need this? It feels like we're trying to be a bit too generic at this point and it's just adding a lot of code branching. Visualize embeddable can wrap its input in `attributes`, and with having a custom save function, things should work fine. Ie.

Reply:
I made this for the case where the _only_ difference between the regular attribute service and the embeddable input that will be using it is the name of the key. This may or may not be the case for Visualize, but I think it's a good feature to support, especially if the 'by value' paradigm will be used this way in many different embeddables. The only new logic this results in is the attribute key line:

https://github.com/elastic/kibana/blob/0249784bd9d0ab74bff7cc0cac3109e2230b6f85/src/plugins/dashboard/public/attribute_service/attribute_service.tsx#L91

Afterwards, the logic is all the same, accessing the attributes via: 

``` typescript
input[this.attributesKey]
```

or setting the attributes via:

``` typescript
{ [this.attributesKey]: newAttributes }
```

Reply:
I actually ended up removing the `attributesKey` to make this PR a little bit simpler. If we experiment with visualize and see that it could be helpful, it would be trivial to re-implement it in a followup PR.